### PR TITLE
Fix NRE in ToolStripItem.InvalidateItemLayout

### LINKED ISSUE #6535: [Regression] NullReferenceException on maximized MDI child icon changes with 150% DPI
* .NET Core Version: 6.0.1

* Have you experienced this same bug with .NET Framework?: No

**Problem description:**

`NullReferenceException` in `System.Windows.Forms.ToolStripItem.InvalidateItemLayout` when you try to change Icon on maximized child form (**only with 150% DPI**):

https://user-images.githubusercontent.com/17767561/150631131-044884bc-7fcf-41c0-9ebb-6f76820ccbc6.mp4

<details>
  <summary>Exception details</summary>
 
  ```
************** Текст исключения **************
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Windows.Forms.ToolStripItem.InvalidateItemLayout(String affectedProperty)
   at System.Windows.Forms.ToolStripItem.set_Image(Image value)
   at System.Windows.Forms.MdiControlStrip.updateIcon()
   at System.Windows.Forms.Form.UpdateWindowIcon(Boolean redrawFrame)
   at System.Windows.Forms.Form.set_Icon(Icon value)
   at WinFormsCoreTest.Form2.button2_Click(Object sender, EventArgs e) in D:\save\projects\core tests\WinFormsCoreTest_MDI_Icon\Form2.cs:line 33
   at System.Windows.Forms.Control.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnClick(EventArgs e)
   at System.Windows.Forms.Button.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)


************** Загруженные сборки **************
System.Private.CoreLib
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Private.CoreLib.dll
----------------------------------------
WinFormsCoreTest
    Версия сборки: 1.0.0.0
    Версия Win32: 1.0.0.0
    CodeBase (база кода): file:///C:/Users/Shapiro/Desktop/net6.0-windows/WinFormsCoreTest.dll
----------------------------------------
System.Runtime
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Runtime.dll
----------------------------------------
System.Windows.Forms
    Версия сборки: 6.0.1.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/System.Windows.Forms.dll
----------------------------------------
System.ComponentModel.Primitives
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.ComponentModel.Primitives.dll
----------------------------------------
System.Windows.Forms.Primitives
    Версия сборки: 6.0.1.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/System.Windows.Forms.Primitives.dll
----------------------------------------
System.Runtime.InteropServices
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Runtime.InteropServices.dll
----------------------------------------
System.Drawing.Primitives
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Drawing.Primitives.dll
----------------------------------------
System.Collections.Specialized
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Collections.Specialized.dll
----------------------------------------
System.Threading
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Threading.dll
----------------------------------------
System.Diagnostics.TraceSource
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Diagnostics.TraceSource.dll
----------------------------------------
System.Collections
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Collections.dll
----------------------------------------
System.Drawing.Common
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/System.Drawing.Common.dll
----------------------------------------
Microsoft.Win32.Primitives
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/Microsoft.Win32.Primitives.dll
----------------------------------------
System.ComponentModel.EventBasedAsync
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.ComponentModel.EventBasedAsync.dll
----------------------------------------
System.Threading.Thread
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Threading.Thread.dll
----------------------------------------
Accessibility
    Версия сборки: 4.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/Accessibility.dll
----------------------------------------
System.ComponentModel.TypeConverter
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.ComponentModel.TypeConverter.dll
----------------------------------------
Microsoft.Win32.SystemEvents
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/Microsoft.Win32.SystemEvents.dll
----------------------------------------
System.Collections.Concurrent
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Collections.Concurrent.dll
----------------------------------------
System.ComponentModel
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.ComponentModel.dll
----------------------------------------
System.Numerics.Vectors
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Numerics.Vectors.dll
----------------------------------------
System.Memory
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Memory.dll
----------------------------------------
System.Resources.Extensions
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/System.Resources.Extensions.dll
----------------------------------------
System.Drawing
    Версия сборки: 6.0.1.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/System.Drawing.dll
----------------------------------------
System.ObjectModel
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.ObjectModel.dll
----------------------------------------
System.Linq
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Linq.dll
----------------------------------------
System.Collections.NonGeneric
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Collections.NonGeneric.dll
----------------------------------------
System.Runtime.CompilerServices.Unsafe
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Runtime.CompilerServices.Unsafe.dll
----------------------------------------
System.Runtime.Loader
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Runtime.Loader.dll
----------------------------------------
System.Windows.Forms.resources
    Версия сборки: 6.0.1.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.WindowsDesktop.App/6.0.1/ru/System.Windows.Forms.resources.dll
----------------------------------------
System.Private.Uri
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Private.Uri.dll
----------------------------------------
System.Diagnostics.StackTrace
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Diagnostics.StackTrace.dll
----------------------------------------
System.Reflection.Metadata
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Reflection.Metadata.dll
----------------------------------------
System.Collections.Immutable
    Версия сборки: 6.0.0.0
    Версия Win32: 6.0.121.56705
    CodeBase (база кода): file:///C:/Program%20Files/dotnet/shared/Microsoft.NETCore.App/6.0.1/System.Collections.Immutable.dll
----------------------------------------
  ```
  
</details>

It's obvious that the regression was somehow introduced [here](https://github.com/dotnet/winforms/pull/4650).
Interesting thing to note that despite the exception, the icon changes successfully.

**Expected behavior:**
No Exception.

**Minimal repro:**
[WinFormsCoreTest_MDI_Icon.zip](https://github.com/dotnet/winforms/files/7948411/WinFormsCoreTest_MDI_Icon.zip)

### PR DESCRIPTION
**Pull request by jbhensley**
Fixes #6535

## Proposed changes

PR #4650 introduced a refresh of the `MdiControlStrip` icon when a maximized MDI child form's icon changes. That surfaced a bug ultimately resulting in a `NullReferenceException`. The fix is trivial, with most work spent understanding the root cause and proving the fix to be correct.

The NRE was triggered when a maximized MDI child form's icon changed if the new icon's size caused the `MdiControlStip`'s size to change. The Rube Goldberg-like chain of events went like this:

- The `ToolStripItem` containing the icon would invalidate its layout
- This would trigger a call to `LayoutTransaction.DoLayout` on the item's owner (the `MdiControlStrip`)
- The `MdiControlStrip` would resize
- The area available to the maximized MDI child form would change due to `MdiControlStrip` taking up more/less space
- The MDI child would receive a `WM_SIZE` message
- The MDI child would call `MdiParentInternal.UpdateMdiControlStrip` in `Form.WmSize`
- The MDI parent would dispose the current `MdiControlStip` and create a new one (it does this unconditionally)
- The disposed `MdiControlStrip` would dispose all of its associated `Items` and clear its `Items` property
- Clearing `Items` would set each `ToolStripItem.Owner` to `null`
- The stack would unwind and the `ToolStripItem` would attempt to call `Owner.Invalidate` with `Owner` being a `null` reference

Because the `ToolStripItem` and its former parent `ToolStrip` are both disposed at this point, we can just use a null conditional on `Owner.Invalidate`. 

I had considered checking `ToolStripItem.IsDisposed` as a possible alternative fix, but that just indirectly verifies the condition we are concerned about. `ToolStripItem.IsDisposed` will be `true` and `Owner` will be `null`. Might as well check `Owner` directly since that's the thing that will trigger the NRE.

## Customer Impact

Customer should not experience an NRE when changing the icon of a maximized MDI child when doing so results in a size change of the `MdiControlStrip` containing it

## Regression? 

- Yes 

## Risk

- Minimal

## Test methodology 

- Unit test added. Fails pre-fix and passes post-fix

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6588)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 1]
```diff
@@ -2328,7 +2328,11 @@ internal void InvalidateItemLayout(string affectedProperty)
             if (Owner is not null)
             {
                 LayoutTransaction.DoLayout(Owner, this, affectedProperty);
-                Owner.Invalidate();
+                /* DoLayout may cause the ToolStrip size to change. If the ToolStrip is an MdiControlStrip, the
+                 * active Mdi child is maximized, and DoLayout causes the size to change then Form.MdiControlStrip
+                 * (Owner) will be disposed and replaced with a new one. This means the current ToolStripItem will
+                 * also be disposed and have no owner on the next line. See issue #6535 */
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs:2334]:
```suggestion
                // DoLayout may cause the ToolStrip size to change. If the ToolStrip is an MdiControlStrip, the
                // active Mdi child is maximized, and DoLayout causes the size to change then Form.MdiControlStrip
                // (Owner) will be disposed and replaced with a new one. This means the current ToolStripItem will
                // also be disposed and have no owner on the next line. See https://github.com/dotnet/winforms/issues/6535.
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MdiControlStripTests.cs [hunk 2]
```diff
@@ -310,6 +310,29 @@ public void MdiControlStrip_MaximizedChildWindow_ControlbBoxButtons_AreNotCloned
...
+            mdiParent.MainMenuStrip = menuStrip;
+
+            mdiParent.Show();
+            mdiChild.Show();
+            mdiChild.WindowState = FormWindowState.Maximized;
+
+            Reflection.PropertyInfo mdiControlStripProperty = typeof(Form).GetProperty("MdiControlStrip", Reflection.BindingFlags.NonPublic | Reflection.BindingFlags.Instance);
+            MdiControlStrip originalMdiControlStrip = (MdiControlStrip)mdiControlStripProperty.GetValue(mdiParent);
+
+            mdiChild.Icon = Icon.FromHandle(new Bitmap(256, 256).GetHicon()); // Force size change with large icon
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MdiControlStripTests.cs:330]: We're leaking a handle here, see TaskDialogIcon.BitmapToIcon for correct implementation.

**Review by RussKie — COMMENTED** [review]
👍

**Comment by RussKie** [comment]
@merriemcgaw @dreddy-work we'll need to service this.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MdiControlStripTests.cs [hunk 3]
```diff
@@ -310,6 +310,33 @@ public void MdiControlStrip_MaximizedChildWindow_ControlbBoxButtons_AreNotCloned
...
+
+            mdiParent.Controls.Add(menuStrip);
+            mdiParent.MainMenuStrip = menuStrip;
+
+            mdiParent.Show();
+            mdiChild.Show();
+            mdiChild.WindowState = FormWindowState.Maximized;
+
+            Reflection.PropertyInfo mdiControlStripProperty = typeof(Form).GetProperty("MdiControlStrip", Reflection.BindingFlags.NonPublic | Reflection.BindingFlags.Instance);
+            MdiControlStrip originalMdiControlStrip = (MdiControlStrip)mdiControlStripProperty.GetValue(mdiParent);
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MdiControlStripTests.cs:328]: I must've missed it during the previous review. We avoid using explicit reflection, and use [`TestAccessor`](https://github.com/dotnet/winforms/search?q=TestAccessor) instead.

Comment (jbhensley) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/MdiControlStripTests.cs]: An object wrapper that uses reflection to expose private properties/methods as `dynamic` ... very cool.

The intention of checking `MdiControlStrip` is just to prove that the test is valid. If it didn't recreate then we didn't trip the condition that led to the NRE. I would rather not reach into an object's internals, but don't see another way here.

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Thank you @jbhensley!

**Pull request merged**